Introduce fine-grained FDSN client exceptions

yetinam opened the pull request:
<!--

Thank your for contributing to ObsPy!

!! Please check that you select the **correct base branch** (details see below link) !!

Before submitting a PR, please review the pull request guidelines:
https://github.com/obspy/obspy/blob/master/CONTRIBUTING.md#submitting-a-pull-request

Also, please make sure you are following the ObsPy branching model:
https://github.com/obspy/obspy/wiki/ObsPy-Git-Branching-Model

-->

### What does this PR do?

This PR introduces fine-grained exceptions for the FDSN client to allow easier exception handling. As all newly introduced exceptions subclass `FDSNException` this change should be compatible with any existing code.

### Why was it initiated?  Any relevant Issues?

Feature was proposed in #2653

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"
- [x] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [ ] All tests still pass.
- [ ] Any new features or fixed regressions are be covered via new tests.
- [ ] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .

yetinam commented:
+TESTS:clients.fdsn

yetinam commented:
There are currently no tests for `FDSNServerException` (triggered if the server returns http code 500) and `FDSNTooManyRequestsException`, as I was not able to consistently trigger them from a web service. I'd appreciate any suggestions on how to test this.

krischer commented:
Instead of relying on webservices to return a certain status (which might change in the future) it would be easier and more stable to mock out the response to trigger all the code paths in the tests. A simplistic way of doing this would be to just mock out the `download_url()` function: https://github.com/obspy/obspy/blob/master/obspy/clients/fdsn/client.py#L1751

Please have a look at the mock library here in case you are not familiar with it: https://docs.python.org/3/library/unittest.mock.html

yetinam commented:
Thanks for the pointer. I'll modify and extend the tests accordingly.

yetinam commented:
I changed the tests so that they uses mocks now and extended them. The test `test_eida_token_resolution` currently fails, because the EIDA token expired.

megies commented:
circle ci fails are unrelated and fixed on master

megies commented:
@krischer you wanna give it the final nod, if you're happy with how the tests are set up?

krischer commented on obspy/clients/fdsn/tests/test_client.py:
```diff
@@ -1448,15 +1456,96 @@ def test_get_waveforms_empty_seed_codes(self):
+
+        download_url_mock.return_value = (400, io.BytesIO(b"token is expired"))
+        self.assertRaises(FDSNAuthenticationException,
+                          self.client.get_stations)
```
Even though this is indeed an authentication related problem the return code is still 400 and I think it would be a bit cleaner if every return code is mapped to a unique exception.

megies commented:
> It might be nice to have the HTTP code as part of the exception classes

requests names it `status_code`, might be worth a thought following that attribute naming maybe?

krischer commented:
> requests names it `status_code`, might be worth a thought following that attribute naming maybe?

Good point!

yetinam commented:
Thanks for the comments. I modified the exceptions to make sure there is a 1:1 mapping between exceptions and HTTP status codes. I also added the `status_code` to the FDSNException.

I did not create a custom exception for code 414, as it currently is a `NotImplementedException`. As this suggests that it might be implemented sometime, I'd suggest keeping this exception here, but I'm fine with changing it as well.

krischer commented:
Thanks a for this! Looks good to me know and can be merged from my side if CI passes.

You'll have to rebase on the latest `master` branch to get rid of the conflict.

yetinam commented:
I rebased the branch.

megies commented:
not sure why travis didn't run.. would be good to see Travis, otherwise FDSN tests are not run

megies commented:
@yetinam can you do something like `git commit --amend --no-edit` and force push, to try to trigger Travis again? I don't think there is a way to manually trigger a Travis build for PRs from branches on forks.

yetinam commented:
That seems to have done the job.

megies commented:
Tests are in, looks like thats only the old fails we had before, but maybe you can look through them real quick if its what you expect from the PR:
http://tests.obspy.org/111494/

yetinam commented:
I agree, looks like the failing tests are unrelated

megies commented:
Thanks @yetinam!
